cluster_config.go

```diff
@@ -409,32 +411,33 @@ func clusterSecretToKey(secret []byte) (string, error) {
 	return key.String(), nil
 }
 
-// BackupState backs up a state according to this configuration's options
-//func (cfg *Config) BackupState(state state.State) error {
-//	if cfg.BaseDir == "" {
-// 		msg := "ClusterConfig BaseDir unset. Skipping backup"
-// 		logger.Warning(msg)
-// 		return errors.New(msg)
-// 	}
-
-// 	folder := filepath.Join(cfg.BaseDir, "backups")
-// 	err := os.MkdirAll(folder, 0700)
-// 	if err != nil {
-// 		logger.Error(err)
-// 		logger.Error("skipping backup")
-// 		return errors.New("skipping backup")
-// 	}
-// 	fname := time.Now().UTC().Format("20060102_15:04:05")
-// 	f, err := os.Create(filepath.Join(folder, fname))
-// 	if err != nil {
-// 		logger.Error(err)
-// 		return err
-// 	}
-// 	defer f.Close()
-// 	err = state.Snapshot(f)
-// 	if err != nil {
-// 		logger.Error(err)
-// 		return err
-// 	}
-// 	return nil
-// }
+// ExportState saves a json representation of a state according to this
+// configuration's options
+func (cfg *Config) ExportState(state state.State) error {
+	if cfg.BaseDir == "" {
+		msg := "ClusterConfig BaseDir unset. Skipping backup"
+		logger.Warning(msg)
+		return errors.New(msg)
+	}
+
+	folder := filepath.Join(cfg.BaseDir, "exports")
+	err := os.MkdirAll(folder, 0700)
+	if err != nil {
+		logger.Error(err)
+		logger.Error("skipping export")
+		return errors.New("skipping export")
+	}
+	fname := time.Now().UTC().Format("20060102_15:04:05")
+	f, err := os.Create(filepath.Join(folder, fname))
+	if err != nil {
+		logger.Error(err)
+		return err
+	}
+	defer f.Close()
+	err = state.Snapshot(f)
```

Comment:
I got very confused with the `Snapshot` name function. It's not the same format as Raft snapshots.

We may want to just use the result of `mapstate.List()` (which each item in the list converted to PinSerial and then everything to JSON). And with that drop the Snapshot method. The Restore method should probably just be used for Migrations (perhaps renamed to Migrate() and without doing json-unmarshaling).

The import would be then a for-loop of `Add()` operations (which can be implemented as `MultiAdd([]api.Pin)`).

I think this would allow to export a state and re-import it to a different `state` implementation which is something nice to have. JSON marshaling and unmarshaling can be ipfs-cluster-service task.